src/firmware/main/app/primitives/shoot_primitive.c

```diff
@@ -0,0 +1,173 @@
+#include "firmware/main/app/primitives/shoot_primitive.h"
+
+#include <math.h>
+#include <stdio.h>
+#include <stdlib.h>
+
+#include "firmware/main/app/control/bangbang.h"
+#include "firmware/main/app/control/control.h"
+#include "firmware/main/app/control/physbot.h"
+#include "firmware/main/shared/physics.h"
+#include "firmware/main/shared/util.h"
+
+// these are set to decouple the 3 axis from each other
+// the idea is to clamp the maximum velocity and acceleration
+// so that the axes would never have to compete for resources
+#define TIME_HORIZON 0.05f  // s
+
+typedef struct ShootPrimitiveState
+{
+    float destination[3], major_vec[2], minor_vec[2], total_rot;
+} ShootPrimitiveState_t;
+DEFINE_PRIMITIVE_STATE_CREATE_AND_DESTROY_FUNCTIONS(ShootPrimitiveState_t)
+
+
+/**
+ * Scales the major acceleration by the distance from the major axis and the
+ * amount required left to rotate. Total rotation and the distance vector should
+ * not be zero so as to avoid divide by zero errors.
+ *
+ * @param pb The PhysBot data container that contains information about the
+ * major and minor axis.
+ * @return void
+ */
+void scale(PhysBot *pb)
+{
+    float maj_disp        = fabsf(pb->maj.disp) - ROBOT_RADIUS;
+    float distance_vector = sqrtf(powf(maj_disp, 2) + powf(pb->min.disp, 2));
+    if (distance_vector != 0)
+    {
+        float abs_factor        = fabsf(pb->min.disp) / distance_vector;
+        float minor_axis_factor = 1 - abs_factor;
+        pb->maj.accel *= minor_axis_factor;
+    }
+
+    // Doesn't work that well. Possibly because total_rot is updated too often
+    //    if (total_rot != 0) {
+    //        float rot_factor = 1 - fabsf(pb->rot.disp / total_rot);
+    //        pb->maj.accel *= rot_factor;
+    //    }
```

Comment:
Commented out code

Remove?

Reply:
There's a bunch of this in many primitives, calling out-of-scope for now.